File: internal/reconcile/csv.go
package reconcile

import (
	"encoding/csv"
	"fmt"
	"io"
	"strings"
	"time"
)

func ReadExpectedCSV(reader io.Reader) ([]ExpectedRecord, error) {
	r := csv.NewReader(reader)
	r.TrimLeadingSpace = true
	head, err := r.Read()
	if err != nil {
		return nil, fmt.Errorf("read CSV header: %w", err)
	}
	columns := make(map[string]int, len(head))
	for i, value := range head {
		columns[strings.ToLower(strings.TrimSpace(value))] = i
	}
	for _, required := range []string{"reference", "direction", "amount", "counterparty"} {
		if _, ok := columns[required]; !ok {
			return nil, fmt.Errorf("missing required CSV column %q", required)
		}
	}

	var records []ExpectedRecord
	seenReferences := make(map[string]int)
	for rowNumber := 2; ; rowNumber++ {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read CSV row %d: %w", rowNumber, err)
		}
		get := func(name string) string {
			i, ok := columns[name]
			if !ok || i >= len(row) {
				return ""
			}
			return strings.TrimSpace(row[i])
		}

		reference := get("reference")
		if reference == "" {
			return nil, fmt.Errorf("row %d: reference is required", rowNumber)
		}
		referenceKey := strings.ToLower(reference)
		if previousRow, exists := seenReferences[referenceKey]; exists {
			return nil, fmt.Errorf("row %d: reference %q duplicates row %d", rowNumber, reference, previousRow)
		}
		seenReferences[referenceKey] = rowNumber
		direction := Direction(strings.ToLower(get("direction")))
		if direction != DirectionInbound && direction != DirectionOutbound && direction != DirectionSelf {
			return nil, fmt.Errorf("row %d: direction must be inbound, outbound, or self", rowNumber)
		}
		amount, err := ParseUSDC(get("amount"))
		if err != nil {
			return nil, fmt.Errorf("row %d: %w", rowNumber, err)
		}
		counterparty := NormalizeAddress(get("counterparty"))
		if !ValidAddress(counterparty) {
			return nil, fmt.Errorf("row %d: counterparty is not a valid EVM address", rowNumber)
		}

		var dueDate *time.Time
		if raw := get("due_date"); raw != "" {
			parsed, err := time.Parse("2006-01-02", raw)
			if err != nil {
				return nil, fmt.Errorf("row %d: due_date must use YYYY-MM-DD", rowNumber)
			}
			dueDate = &parsed
		}

		records = append(records, ExpectedRecord{
			Reference:    reference,
			Direction:    direction,
			AmountMicros: amount,
			Counterparty: counterparty,
			DueDate:      dueDate,
			MemoID:       strings.ToLower(get("memo_id")),
		})
	}
	if len(records) == 0 {
		return nil, fmt.Errorf("CSV contains no expected records")
	}
	return records, nil
}

File: web/src/csv.ts
import type { Direction, ExpectedInput } from "./types";

const requiredColumns = ["reference", "direction", "amount", "counterparty"] as const;
const addressPattern = /^0x[0-9a-fA-F]{40}$/;

export function parseExpectedCSV(source: string): ExpectedInput[] {
  const rows = parseRows(source);
  if (rows.length === 0) throw new Error("The CSV is empty.");
  const header = rows[0].map((value) => value.trim().toLowerCase());
  const indexes = new Map(header.map((value, index) => [value, index]));
  for (const column of requiredColumns) {
    if (!indexes.has(column)) throw new Error(`Missing required column: ${column}`);
  }

  const records: ExpectedInput[] = [];
  const references = new Set<string>();
  for (let rowIndex = 1; rowIndex < rows.length; rowIndex += 1) {
    const row = rows[rowIndex];
    if (row.every((value) => value.trim() === "")) continue;
    const get = (column: string) => {
      const index = indexes.get(column);
      return index === undefined ? "" : (row[index] ?? "").trim();
    };
    const reference = get("reference");
    if (!reference) throw new Error(`Row ${rowIndex + 1}: reference is required.`);
    const referenceKey = reference.toLowerCase();
    if (references.has(referenceKey)) {
      throw new Error(`Row ${rowIndex + 1}: reference ${reference} is duplicated.`);
    }
    references.add(referenceKey);

    const direction = get("direction") as Direction;
    if (!(["inbound", "outbound", "self"] as string[]).includes(direction)) {
      throw new Error(`Row ${rowIndex + 1}: direction must be inbound, outbound, or self.`);
    }
    const amount = get("amount");
    if (!/^\d+(\.\d{1,6})?$/.test(amount) || Number(amount) <= 0) {
      throw new Error(`Row ${rowIndex + 1}: amount must be positive with at most 6 decimals.`);
    }
    const counterparty = get("counterparty");
    if (!addressPattern.test(counterparty)) {
      throw new Error(`Row ${rowIndex + 1}: counterparty is not a valid address.`);
    }
    const dueDate = get("due_date");
    if (dueDate && !/^\d{4}-\d{2}-\d{2}$/.test(dueDate)) {
      throw new Error(`Row ${rowIndex + 1}: due_date must use YYYY-MM-DD.`);
    }
    records.push({
      reference,
      direction,
      amount,
      counterparty: counterparty.toLowerCase(),
      ...(dueDate ? { dueDate } : {}),
      ...(get("memo_id") ? { memoId: get("memo_id").toLowerCase() } : {}),
    });
  }
  if (records.length === 0) throw new Error("The CSV contains no records.");
  return records;
}

function parseRows(source: string): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let quoted = false;
  for (let index = 0; index < source.length; index += 1) {
    const character = source[index];
    if (quoted) {
      if (character === '"' && source[index + 1] === '"') {
        field += '"';
        index += 1;
      } else if (character === '"') {
        quoted = false;
      } else {
        field += character;
      }
      continue;
    }
    if (character === '"' && field === "") quoted = true;
    else if (character === ",") {
      row.push(field);
      field = "";
    } else if (character === "\n") {
      row.push(field.replace(/\r$/, ""));
      rows.push(row);
      row = [];
      field = "";
    } else field += character;
  }
  if (quoted) throw new Error("The CSV contains an unclosed quoted field.");
  if (field !== "" || row.length > 0) {
    row.push(field.replace(/\r$/, ""));
    rows.push(row);
  }
  return rows;
}
